### PR TITLE
Adding support for nats-streaming servers.

### PR DESCRIPTION
**Pull request by JesseObrien**
## Motivation
* Adds support for [nats streaming](https://docs.nats.io/nats-streaming-concepts/intro)

## Modifications
* Addition of a new package, `nats`

## Verify change
* [ ] Make sure the change passes the CI checks.

### CODE REVIEW

File: examples/nats/go.mod [hunk 1]
```diff
@@ -0,0 +1,10 @@
+module nats-example
+
+go 1.14
+
+
+require (
+  github.com/reugn/go-streams/nats v0.0.0
+)
+
+replace github.com/reugn/go-streams/nats => ../../nats
```
Comment (reugn) [inline examples/nats/go.mod:10]: It's better to require the NATS connector in the go.mod file under the examples folder with all others.

File: examples/nats/go.sum [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+github.com/reugn/go-streams v0.6.3/go.mod h1:OsEa/+BuietQY5sJUCm5nrZujfWtuKIv7KQVCZCXYGg=
+golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2/go.mod h1:djNgcEr1/C05ACkg1iLfiJU5Ep61QUkGW8qpdssI0+w=
+golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4/go.mod h1:yigFU9vqHzYiE8UmvKecakEJjdnWj3jj499lnFckfCI=
+golang.org/x/crypto v0.0.0-20200323165209-0ec3e9974c59 h1:3zb4D3T4G8jdExgVU/95+vQXfpEPiMdCaZgmGVxjNHM=
+golang.org/x/crypto v0.0.0-20200323165209-0ec3e9974c59/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
+golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
+golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
+golang.org/x/sys v0.0.0-20190412213103-97732733099d/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
+golang.org/x/tools v0.0.0-20181030221726-6c7e314b6563/go.mod h1:n7NCudcB/nEzxVGmLbDWY5pfWTLqBcC2KZ6jyYvM4mQ=
```
Comment (reugn) [inline examples/nats/go.sum:26]: Should relocate with the go.mod file.

File: go.mod [hunk 3]
```diff
@@ -1,3 +1,5 @@
 module github.com/reugn/go-streams
 
 go 1.15
+
+require github.com/nats-io/stan.go v0.7.0 // indirect
```
Comment (reugn) [inline go.mod:5]: NATS deps should go under it's own go.mod file.

File: go.sum [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+github.com/nats-io/stan.go v0.7.0/go.mod h1:Ci6mUIpGQTjl++MqK2XzkWI/0vF+Bl72uScx7ejSYmU=
+golang.org/x/crypto v0.0.0-20190308221718-c2843e01d9a2/go.mod h1:djNgcEr1/C05ACkg1iLfiJU5Ep61QUkGW8qpdssI0+w=
+golang.org/x/crypto v0.0.0-20190701094942-4def268fd1a4/go.mod h1:yigFU9vqHzYiE8UmvKecakEJjdnWj3jj499lnFckfCI=
+golang.org/x/crypto v0.0.0-20200323165209-0ec3e9974c59 h1:3zb4D3T4G8jdExgVU/95+vQXfpEPiMdCaZgmGVxjNHM=
+golang.org/x/crypto v0.0.0-20200323165209-0ec3e9974c59/go.mod h1:LzIPMQfyMNhhGPhUkYOs5KpL4U8rLKemX1yGLhDgUto=
+golang.org/x/net v0.0.0-20190404232315-eb5bcb51f2a3/go.mod h1:t9HGtf8HONx5eT2rtn7q6eTqICYqUVnKs3thJo3Qplg=
+golang.org/x/sys v0.0.0-20190215142949-d0b11bdaac8a/go.mod h1:STP8DvDyc/dI5b8T5hshtkjS+E42TnysNCUPdjciGhY=
+golang.org/x/sys v0.0.0-20190412213103-97732733099d/go.mod h1:h1NjWce9XRLGQEsW7wpKNCjG9DtNlClVuFLEZdDNbEs=
+golang.org/x/text v0.3.0/go.mod h1:NqM8EUOU14njkJ3fqMW+pc6Ldnwhi/IjpwHt7yyuwOQ=
+golang.org/x/tools v0.0.0-20181030221726-6c7e314b6563/go.mod h1:n7NCudcB/nEzxVGmLbDWY5pfWTLqBcC2KZ6jyYvM4mQ=
```
Comment (reugn) [inline go.sum:24]: NATS conector deps.

File: nats/nats_stan.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+	return &src
+}
+
+func (ns *NatsSource) init() {
+	sigchan := make(chan os.Signal, 1)
+	signal.Notify(sigchan, syscall.SIGINT, syscall.SIGTERM)
+
+	// Bind all topic subscribers
+	for _, topic := range ns.topics {
+		go func() {
```
Comment (reugn) [inline nats/nats_stan.go:58]: Pass topic as a function argument.

**Review by reugn — CHANGES_REQUESTED** [review]
@JesseObrien thanks for the PR! Please find my review below.

File: examples/nats/go.sum (hunk 2 above)
Comment (JesseObrien) [reply examples/nats/go.sum]: 💯

File: go.mod (hunk 3 above)
Comment (JesseObrien) [reply go.mod]: Yes absolutely. Not sure how these got into the root of the project.

File: examples/nats/go.mod (hunk 1 above)
Comment (JesseObrien) [reply examples/nats/go.mod]: Nice one, I missed those at the examples folder root.

**Comment by JesseObrien** [comment]
Appreciate that review. Should have addressed all the go module weirdness 👍

**Comment by reugn** [comment]
Looking good! One last thing, could you please revert the changes to the main go.mod and go.sum files.

**Comment by JesseObrien** [comment]
Done and done! I didn't mean to delete them 🤦

**Review by reugn — APPROVED** [review]

**Pull request merged**